### PR TITLE
check for ingressIP when ingressProfiles > 1

### PR DESCRIPTION
**Pull request by karanmagdani**
Signed-off-by: Karan.Magdani <kmagdani@redhat.com>

### Which issue this PR addresses:

Fixes https://github.com/Azure/ARO-RP/issues/2010
Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/13948245

### What this PR does / why we need it:

Bug where we end up selecting wrong ingress when there are multiple ingressProfiles.

### CODE REVIEW

File: pkg/operator/deploy/deploy.go [hunk 1]
```diff
@@ -138,6 +138,15 @@ func (o *operator) resources() ([]kruntime.Object, error) {
 		domain += "." + o.env.Domain()
 	}
 
+	ingressIP := o.oc.Properties.IngressProfiles[0].IP
+	if o.oc.Properties.IngressProfiles != nil && len(o.oc.Properties.IngressProfiles) > 1 {
```
Comment (s-amann) [inline pkg/operator/deploy/deploy.go:142]: At this point we would have already blown up if `o.oc.Properties.IngressProfiles` was nil, perhaps we should invert the condition and return an error if it is nil and then proceed with the logic of using the first one or the one named `router-default`

File: pkg/operator/deploy/deploy.go [hunk 2]
```diff
@@ -138,6 +138,15 @@ func (o *operator) resources() ([]kruntime.Object, error) {
...
+		for _, p := range o.oc.Properties.IngressProfiles {
+			if p.Name == "router-default" {
```
Comment (bennerv) [inline pkg/operator/deploy/deploy.go:144]: I think the name is `default` not `router-default`

```json
    "ingressProfiles": [
      {
        "ip": "1.2.3.4",
        "name": "default",
        "visibility": "Public"
      }
    ],
```

File: pkg/operator/deploy/deploy.go (hunk 1 above)
Comment (karanmagdani) [reply pkg/operator/deploy/deploy.go]: You're right lol! thanks, fixed

File: pkg/operator/deploy/deploy.go (hunk 2 above)
Comment (karanmagdani) [reply pkg/operator/deploy/deploy.go]: Fixed, Thanks!!

**Review by s-amann — DISMISSED** [review]

**Review by bryanro92 — COMMENTED** [review]
lgtm, but I wonder if we can use this as an opportunity to write unit test for `deploy.go`. even if the only coverage would be for this added block of code which could easily be refactored into a function

File: pkg/operator/deploy/deploy.go [hunk 3]
```diff
@@ -138,6 +138,15 @@ func (o *operator) resources() ([]kruntime.Object, error) {
 		domain += "." + o.env.Domain()
 	}
 
+	ingressIP := o.oc.Properties.IngressProfiles[0].IP
```
Comment (m1kola) [inline pkg/operator/deploy/deploy.go:141]: `o.oc.Properties.IngressProfiles` can be `nil`, I think.

File: pkg/operator/deploy/deploy.go [hunk 4]
```diff
@@ -138,6 +138,15 @@ func (o *operator) resources() ([]kruntime.Object, error) {
...
+	if len(o.oc.Properties.IngressProfiles) > 1 {
+		for _, p := range o.oc.Properties.IngressProfiles {
+			if p.Name == "default" {
+				ingressIP = p.IP
+			}
```
Comment (m1kola) [inline pkg/operator/deploy/deploy.go:146]: Ingress profile with name `"default"` is not guaranteed to exist. I think customers can create an ingress controller `"foo"` and delete `"default"`. What happens in this case?

**Review by m1kola — DISMISSED** [review]

File: pkg/operator/deploy/deploy.go (hunk 4 above)
Comment (bryanro92) [reply pkg/operator/deploy/deploy.go:146]: @karanmagdani can we combine my suggestion of adding some unit testing here with @m1kola's points about edge cases and make sure we are properly handling all cases?

Comment (karanmagdani) [reply pkg/operator/deploy/deploy.go]: Added test cases and covered in function. 
For case when: there is no default, it just picks the first ingress profile available (what we currently do). IMO this is fine because i don't think we have a way to find the "non-default" default

File: pkg/operator/deploy/deploy.go (hunk 3 above)
Comment (karanmagdani) [reply pkg/operator/deploy/deploy.go]: Fixed, Thanks

File: pkg/operator/deploy/deploy.go [hunk 5]
```diff
@@ -314,6 +319,23 @@ func (o *operator) IsReady(ctx context.Context) (bool, error) {
...
+func GetDefaultImageIP(ingressProfiles []api.IngressProfile) (string, error) {
+	if ingressProfiles == nil || len(ingressProfiles) < 1 {
+		return "", fmt.Errorf("No Ingress Profiles found")
+	}
+	ingressIP := ingressProfiles[0].IP
+	if len(ingressProfiles) > 1 {
+		for _, p := range ingressProfiles {
+			if p.Name == "default" {
+				ingressIP = p.IP
+				return ingressIP, nil
```
Comment (s-amann) [inline pkg/operator/deploy/deploy.go:332]: nit: we can skip assignment and return `p.IP`
```suggestion
				return p.IP, nil
```

File: pkg/operator/deploy/deploy_test.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+				}
+			},
+			want:    "1.1.1.1",
+			wantErr: ``,
+		},
+		{
+			name: "No IngressProfiles, Error",
+			oc: func() *api.OpenShiftClusterProperties {
+				return &api.OpenShiftClusterProperties{
+					IngressProfiles: []api.IngressProfile{},
```
Comment (s-amann) [inline pkg/operator/deploy/deploy_test.go:99]: nit: should be super easy to add the final case of `nil` `IngressProfiles` :)

**Review by s-amann — APPROVED** [review]
lgtm, my nits are just nits :)

File: pkg/operator/deploy/deploy.go (hunk 5 above)
Comment (karanmagdani) [reply pkg/operator/deploy/deploy.go:332]: Good point, thanks!

File: pkg/operator/deploy/deploy_test.go (hunk 6 above)
Comment (karanmagdani) [reply pkg/operator/deploy/deploy_test.go:99]: fixed, thank you

File: pkg/operator/deploy/deploy.go (hunk 5 above)
Comment (karanmagdani) [reply pkg/operator/deploy/deploy.go]: fixed, thank you!

**Review by s-amann — APPROVED** [review]
lgtm

File: pkg/operator/deploy/deploy.go [hunk 7]
```diff
@@ -138,6 +138,11 @@ func (o *operator) resources() ([]kruntime.Object, error) {
 		domain += "." + o.env.Domain()
 	}
 
+	ingressIP, err := GetDefaultImageIP(o.oc.Properties.IngressProfiles)
```
Comment (bryanro92) [inline pkg/operator/deploy/deploy.go:141]: Why is this function called GetDefaultImageIP? Shouldn't it be IngressIP?

**Review by bryanro92 — CHANGES_REQUESTED** [review]
I think function has confusing name

**Review by s-amann — APPROVED** [review]
lgtm

File: pkg/operator/deploy/deploy.go [hunk 8]
```diff
@@ -314,6 +319,22 @@ func (o *operator) IsReady(ctx context.Context) (bool, error) {
 	return true, nil
 }
 
+// o.oc.Properties.IngressProfile
+func GetIngressIP(ingressProfiles []api.IngressProfile) (string, error) {
```
Comment (cadenmarchese) [inline pkg/operator/deploy/deploy.go:323]:
```suggestion
func getIngressIP(ingressProfiles []api.IngressProfile) (string, error) {
```

File: pkg/operator/deploy/deploy_test.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+// Licensed under the Apache License 2.0.
+
+import (
+	"testing"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/cmp"
+)
+
+func TestIngressProfiles(t *testing.T) {
```
Comment (cadenmarchese) [inline pkg/operator/deploy/deploy_test.go:13]: Maybe this naming makes more sense?
```suggestion
func TestGetIngressIP(t *testing.T) {
```

File: pkg/operator/deploy/deploy_test.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+	type test struct {
+		name    string
+		oc      func() *api.OpenShiftClusterProperties
+		want    string
+		wantErr string
```
Comment (cadenmarchese) [inline pkg/operator/deploy/deploy_test.go:19]:
```suggestion
		wantErr error
```

File: pkg/operator/deploy/deploy_test.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+					IngressProfiles: []api.IngressProfile{
+						{
+							Name: "default",
+							IP:   "1.2.3.4",
+						},
+					},
+				}
+			},
+			want:    "1.2.3.4",
+			wantErr: ``,
```
Comment (cadenmarchese) [inline pkg/operator/deploy/deploy_test.go:36]:
```suggestion
			wantErr: nil,
```
Same suggestion for below cases - your function returns `error` type so we should probably use the same type in the test.

File: pkg/operator/deploy/deploy_test.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+			want:    "",
+			wantErr: "No Ingress Profiles found",
+		},
+		{
+			name: "Nil IngressProfiles, Error",
+			oc: func() *api.OpenShiftClusterProperties {
+				return &api.OpenShiftClusterProperties{}
+			},
+			want:    "",
+			wantErr: "No Ingress Profiles found",
```
Comment (cadenmarchese) [inline pkg/operator/deploy/deploy_test.go:111]:
```suggestion
			wantErr: errors.New("No Ingress Profiles found"),
```

File: pkg/operator/deploy/deploy.go [hunk 13]
```diff
@@ -138,6 +138,11 @@ func (o *operator) resources() ([]kruntime.Object, error) {
 		domain += "." + o.env.Domain()
 	}
 
+	ingressIP, err := GetIngressIP(o.oc.Properties.IngressProfiles)
```
Comment (cadenmarchese) [inline pkg/operator/deploy/deploy.go:141]:
```suggestion
	ingressIP, err := getIngressIP(o.oc.Properties.IngressProfiles)
```

**Review by cadenmarchese — CHANGES_REQUESTED** [review]
Nice work on this, just some minor suggestions / nits for consistency.

File: pkg/operator/deploy/deploy.go (hunk 8 above)
Comment (karanmagdani) [reply pkg/operator/deploy/deploy.go]: Fixed

File: pkg/operator/deploy/deploy_test.go (hunk 12 above)
Comment (karanmagdani) [reply pkg/operator/deploy/deploy_test.go]: Fixed

File: pkg/operator/deploy/deploy.go (hunk 13 above)
Comment (karanmagdani) [reply pkg/operator/deploy/deploy.go]: Thank you! Fixed as per suggestions

**Review by s-amann — APPROVED** [review]

**Review by s-amann — DISMISSED** [review]

**Review by bryanro92 — CHANGES_REQUESTED** [review]
nit: name of function

File: pkg/operator/deploy/deploy.go [hunk 14]
```diff
@@ -314,6 +320,21 @@ func (o *operator) IsReady(ctx context.Context) (bool, error) {
 	return true, nil
 }
 
+func getIngressIP(ingressProfiles []api.IngressProfile) (string, error) {
```
[the remark this answers is not in the document]
Comment (karanmagdani) [reply pkg/operator/deploy/deploy.go]: Fixed, changed to checkIngressIP

**Review by bryanro92 — APPROVED** [review]
lgtm, we may need a controller portion of this to watch the ingress profile from within the cluster, but this is at least an iteration in a better direction
